synapse/util/caches/ttlcache.py

```diff
@@ -40,26 +45,27 @@ def __init__(self, cache_name, timer=time.time):
 
         self._metrics = register_cache("ttl", cache_name, self, resizable=False)
 
-    def set(self, key, value, ttl):
+    def set(self, key: KT, value: VT, ttl: float) -> None:
         """Add/update an entry in the cache
 
         Args:
             key: key for this entry
             value: value for this entry
-            ttl (float): TTL for this entry, in seconds
+            ttl: TTL for this entry, in seconds
         """
         expiry = self._timer() + ttl
 
         self.expire()
         e = self._data.pop(key, SENTINEL)
-        if e != SENTINEL:
+        if e is not SENTINEL:
+            assert isinstance(e, _CacheEntry)
             self._expiry_list.remove(e)
 
         entry = _CacheEntry(expiry_time=expiry, ttl=ttl, key=key, value=value)
         self._data[key] = entry
         self._expiry_list.add(entry)
 
-    def get(self, key, default=SENTINEL):
+    def get(self, key: KT, default: T = SENTINEL) -> Union[VT, T]:  # type: ignore
```

Comment:
I could not get mypy to not complain about this:

> synapse/util/caches/ttlcache.py:68: error: Incompatible default for argument "default" (default has type "object", argument has type "T")  [assignment]

We `dictionary_cache` seems to have a magic `_Sentinel` enum to help with this, but I couldn't get it to work.

Reply:
I'm fairly keen to figure this out since we'd want to use the same pattern for all the caches FTR.

Reply:
After some testing I think doing `SENTINEL = object()  # type: Any` works. It's a bit cheeky since `SENTINEL` isn't a `T`, but I think its probably the easiest.